pymbolic/mapper/__init__.py

```diff
@@ -194,6 +195,34 @@ def map_foreign(self, expr, *args, **kwargs):
                     "{} encountered invalid foreign object: {}".format(
                         self.__class__, repr(expr)))
 
+
+class CachedMapper(Mapper):
+    """
+    A mapper that memoizes the mapped result for the expressions traversed.
+
+    .. automethod:: get_cache_key
+    """
+    def __init__(self):
+        self._cache: Dict[Any, Any] = {}
+
+    def get_cache_key(self, expr, *args, **kwargs):
+        """
+        A cache key implementation assuming *expr* is immutable while elements
+        of *args*, *kwargs* aren't.
+        """
+        return ((id(expr), expr), args, tuple(sorted(kwargs.items())))
+
+    def __call__(self, expr, *args, **kwargs):
```

Comment:
Tried to give it a shot, but I think it would be too disruptive for this PR. Essentially for every `Mapper` class, we would need `MapperWithArgs`, `MapperWithArgsKwargs` and there are a fair amount of mappers.